Add RestrictFileSystems= property using LSM BPF

iaguis opened the pull request:
This PR adds the `RestrictFileSystems=` property. When used, processes
belonging to a service are only able to access the filesystems listed in the
property.

This is implemented by attaching a BPF program to the `file_open` BPF LSM hook.
The program is attached at boot time and stays there forever. Then, when a
service specifying the `RestrictFileSystems=` property is started, an entry is
added to a global hash of maps BPF map. The map stores a set of filesystem
magic numbers per cgroupID. When a process tries to open a file, the BPF
program is executed and checks the cgroup the process is running in: if an
entry is present in the global map it checks if the filesystem the process is
trying to access is present in the set, if not, it denies access to it.

`RestrictFileSystems=` is only supported on systems with the LSM BPF hook
enabled and using cgroup2 (unified or hybrid).

This PR makes use of the libbpf framework introduced in #17655. Same as that PR,
it requires clang, llvm and libbpf at compile time, and the
[libbpf](https://github.com/libbpf/libbpf) shared library is optional at runtime (uses dlopen).

Thanks to the usage of libbpf, the program can use the CO-RE (Compile-Once
Run-Everywhere) technology so it doesn't require kernel headers at runtime to
access internal kernel structures.

boucman commented on man/systemd.exec.xml:
```diff
@@ -1699,6 +1699,25 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+
+        <listitem><para>Restricts the set of filesystems processes of this unit can open files on. Takes a space-separated
+        list of filesystem names. Any filesystem listed is made accessible to the unit's processes, access to filesystem
+        types not listed is prohibited (allow-listing). If the empty string is assigned, access to filesystems is not
```
would it be possible to add the negation syntax ? 
The first use case I can think of is denying access to fat/vfat to some processes and I might not know the kind of filesystem the rootfs itself is on..

Alternatlively some some sort of AllowFilesystemsForDirectory= where we give a path and it allows whatever  filesystem that path is on could be usefull

iaguis replied:
> would it be possible to add the negation syntax ?

The current BPF implementation doesn't support this but it shouldn't be too difficult to support. I agree this would be nice.

> Alternatlively some some sort of AllowFilesystemsForDirectory= where we give a path and it allows whatever filesystem that path is on could be usefull

Not sure this is too useful since it will only support one filesystem, the negation syntax is probably more flexible and should cover the same use case although the user would have to specify the filesystem type. We could also make it a list of directories but it feels weird to me.

boucman commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -3210,6 +3212,8 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
     <!--property RestrictNamespaces is not documented!-->
 
+    <!--property RestrictFileSystems is not documented!-->
+
```
Please add documentation for the dbus API too

I know it's new and everything, but let's try to take the good habits from the start

iaguis replied:
Didn't realize this was new, I thought not documenting it was a conscious decision. Will do.

boucman commented on man/systemd.exec.xml:
```diff
@@ -1699,6 +1699,25 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+
+        <para>Note that this setting might not be supported on some systems (for example if the LSM eBPF hook is
+        not enabled in the underlying kernel or if not using the unified control group hierarchy). This setting
+        will fail in that case.</para>
```
Fail as in "deny access" or as in "not implementing the security" 
The general systemd philosophy is that if a feature is not available, things should just work. so we probably want the latter

iaguis replied:
In this case I meant fail as in "the service won't start and will print an error". I chose this because this is a security feature and I didn't want to give a false sense of security. However you're right that in general things just work when they're not supported, I'm on the fence here.

bluca replied:
IIRC all the seccomp directives and seccomp directives are also best effort - if you build without seccomp/selinux support, they are simply ignored. I think the same pattern should be used - doesn't make much sense to diverge

boucman replied:
yes... the other reason to do that is so that the same .service can be used on any distribution/setup. i.e distributions can use upstream provided services without haing to worry what is or isn't supported in their particular kernel

boucman commented on src/basic/stat-util.c:
```diff
@@ -471,3 +471,154 @@ int statx_fallback(int dfd, const char *path, int flags, unsigned mask, struct s
+        else if (streq(name, "zonefs"))
+                *ret = ZONEFS_MAGIC;
+        else if (streq(name, "zsmalloc"))
+                *ret = ZSMALLOC_MAGIC;
```
How would a user find that list of magic names ? is that the same one as the one the "mount" comand use ?
how to make that future-proof wrt futre filesystems or out-of-tree filesystems ? is there a better way to do that if/elsif list ?

iaguis replied:
This is a good point, I've used the names in the `file_system_type.name` struct in the kernel, I assume they're the same ones the "mount" command uses but I'm not sure.

I don't have good ideas on how to make it future proof, I don't think there's any API to get the magic number for a name so it'd have to be something manual.

boucman reviewed (commented):
I'm not a systemd devs, so take all my remarks as suggestions..

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1313,6 +1313,28 @@ int cg_pid_get_machine_name(pid_t pid, char **machine) {
+        assert(path);
+        assert(ret);
+
+        h = malloc0(offsetof(struct file_handle, f_handle) + sizeof(uint64_t));
```
this is fixed size and small afaics, allocate on stack plz

poettering commented on src/basic/stat-util.c:
```diff
@@ -471,3 +471,154 @@ int statx_fallback(int dfd, const char *path, int flags, unsigned mask, struct s
+        assert(name);
+        assert(ret);
+
+        if (streq(name, "apparmorfs"))
```
can't we generate this automatically from linux/magic.h (plus a few manual additions)? or at least have some tool that validates on build that everything in linux/magic.h is also listed here?

also, please use gperf for this, see errno-from-gperf.h and such

iaguis replied:
I'll have a look.

poettering reviewed (commented):
too late today or a full review, but some early comments

poettering commented:
> Note the BPF LSM hook is very new and is not enabled in current distro kernels AFAIK.

Hmm, pity. Is this tracked anywhere? Did anyone ever file a bug against let's say Fedora asking for this to be added yet?

poettering commented:
i finally reviewed @wat-ze-hex's PR #16655 the other day, the commits stemming from there i'll not re-review here.

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,264 @@
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0)
+                return log_debug_errno(errno, "Failed to create BPF map");
```
all other error conditions in this functions result in a LOG_ERR error, this one otoh is LOG_DEBUG.

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,264 @@
+                return log_error_errno(errno, "Error populating LSM BPF map: %m");
+
+        STRV_FOREACH(f, filesystems) {
+                if (DEBUG_LOGGING)
```
drop this conditionalization, it's implicit in `log_debug()`.

(you want to use `if (DEBUG_LOGGING)` if you need to do some expensive work to be able to print the debug message, but this is not the case here, hence just call log_debug() directly)

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,264 @@
+}
+
+int lsm_bpf_setup(void) {
+        return -EOPNOTSUPP;
```
this should probably output an error at LOG_ERR level.

Our coding style regarding error logging is that either a function is of the logging kind or of the non-logging kind. The latter are "library-like" functions where logging is left for the caller to do, the latter are more "main-program-like" functions, where logging is done itself, and the caller shouldn't log anymore. You should decide what kind of function your is (I'd say: main-program-like) and then make sure you log in all cases or in no cases. But in this current version of your patch you log in most error cases at LOG_ERR level, but if supported is not compiled in then you will not log at all...

So, the goal really is to avoid duplicate logging as well as missing logging: we want that every error is logged exactly once, to make things as little confusing as possible.

(Note that in the discussion above LOG_DEBUG level logging doesn't count as real logging... debug logging is always OK, regardless if library-like or main-program-like)

poettering commented on src/core/bpf-lsm.h:
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#define LSM_MAP_BPFFS_PATH "/sys/fs/bpf/systemd/lsm_bpf_map"
```
why is this a public define?

poettering commented on src/core/bpf-lsm.h:
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
```
should have `#pragma once` like all our header files

poettering commented on src/core/cgroup.c:
```diff
@@ -7,6 +7,9 @@
+#if HAVE_LIBBPF
+#include "bpf-lsm.h"
+#endif
 #include "bpf-firewall.h"
```
we try to keep our includes sorted alphabetically

poettering commented on src/core/cgroup.c:
```diff
@@ -7,6 +7,9 @@
 #include "bpf-devices.h"
+#if HAVE_LIBBPF
+#include "bpf-lsm.h"
+#endif
```
could we move this ifdeffery into the header file please? it's usually better to hide ifdeffery in provider not the user of the code if possible

poettering commented on src/core/execute.h:
```diff
@@ -299,6 +299,8 @@ struct ExecContext {
 
         unsigned long restrict_namespaces; /* The CLONE_NEWxyz flags permitted to the unit's processes */
 
+        char **restrict_filesystems;
```
nothing frees this?

poettering commented on src/core/load-fragment.c:
```diff
@@ -3509,6 +3509,73 @@ int config_parse_restrict_namespaces(
+                        break;
+                }
+
+                k = TAKE_PTR(word);
```
hmm? why?

poettering commented on src/core/load-fragment.c:
```diff
@@ -3509,6 +3509,73 @@ int config_parse_restrict_namespaces(
+
+                r = fs_type_from_string(k, &magic_number);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, 0,
```
5th arg should be `r`, i.e. always pass error codes you got on to log_syntax() (it will then appear in structured logging)

poettering commented on src/core/load-fragment.c:
```diff
@@ -3509,6 +3509,73 @@ int config_parse_restrict_namespaces(
+        if (n) {
+                r = strv_extend_strv(restrict_fs, n, true);
+                if (r < 0)
+                        return r;
```
doesn't matter much, but maybe call strv_sort() here too. just in case, so that the introspected result alwas looks the saame.

bluca replied:
@iaguis this comment is still unaddressed

iaguis replied:
Hm, this is now a Set which is not ordered. I'm not sure how to handle this. We sort in property_get_restrict_filesystems(), anything else we should do?

bluca replied:
oh sorry, I had missed that - you can ignore this then

poettering commented on man/systemd.exec.xml:
```diff
@@ -1699,6 +1699,25 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+
+        <para>Note that this setting might not be supported on some systems (for example if the LSM eBPF hook is
+        not enabled in the underlying kernel or if not using the unified control group hierarchy). This setting
+        will fail in that case.</para></listitem>
```
i figure sooner or later we need some kind of grouping here. similar to the allowlisting we have for syscalls. i.e. `@basic-api` for procfs + sysfs, then maybe `@common-block` for ext4, xfs, btrfs; `@network` for well-known network file systems, and so on

poettering replied:
(the `systemd-analyze filesystems` command i proposed elsewhere might be the right tool to show these groups)

poettering replied:
And of course  is_network_fs(),  is_temporary_fs(), fstype_is_network(), fstype_is_api_vfs() should be updated to use the same database...

poettering commented on src/test/test-bpf-lsm.c:
```diff
@@ -0,0 +1,108 @@
+                        return log_unit_error_errno(u, r, "Failed to parse RestrictFileSystems: %m");
+        }
+
+        exec_start = strjoin("cat ", file_path);
```
assert_se() around this please, to catch oom conditions

poettering commented on src/test/test-bpf-lsm.c:
```diff
@@ -0,0 +1,108 @@
+
+        cld_code = SERVICE(u)->exec_command[SERVICE_EXEC_START]->exec_status.code;
+        if (cld_code != CLD_EXITED) {
+                r = -SYNTHETIC_ERRNO(EBUSY);
```
SYNTHETIC_ERRNO() exists purely to pass some extra info to log_error_errno(), it should only be used around the first arg of log_error_errno() (and related calls), but never really assigned to anything else. please move this into the first arg of the function call hence.

poettering commented on src/test/test-bpf-lsm.c:
```diff
@@ -0,0 +1,108 @@
+        }
+
+        if (SERVICE(u)->state != SERVICE_DEAD) {
+                r = -SYNTHETIC_ERRNO(EBUSY);
```
ditto

poettering commented on src/test/test-bpf-lsm.c:
```diff
@@ -0,0 +1,108 @@
+
+        assert_se(getrlimit(RLIMIT_MEMLOCK, &rl) >= 0);
+        rl.rlim_cur = rl.rlim_max = MAX(rl.rlim_max, CAN_MEMLOCK_SIZE);
+        (void) setrlimit(RLIMIT_MEMLOCK, &rl);
```
setrlimit_closest()

poettering commented on src/test/test-bpf-lsm.c:
```diff
@@ -0,0 +1,108 @@
+        assert_se(test_restrict_filesystems(m, "restrict_filesystems_test.service", "/sys/kernel/debug/sleep_time", STRV_MAKE("btrfs", "ext4")) < 0);
+        assert_se(test_restrict_filesystems(m, "restrict_filesystems_test.service", "/sys/kernel/debug/sleep_time", STRV_MAKE("debugfs", "btrfs", "ext4")) >= 0);
+
+        return 0;
```
i figure most of the comments I wrote to @wat-ze-hex' test PR also apply here

poettering reviewed (changes requested):
great work!

would love @wat-ze-hex input/review on this one!

poettering commented:
also needs rebase

bluca commented:
> > Note the BPF LSM hook is very new and is not enabled in current distro kernels AFAIK.
> 
> Hmm, pity. Is this tracked anywhere? Did anyone ever file a bug against let's say Fedora asking for this to be added yet?

I'll start with Debian: https://salsa.debian.org/kernel-team/linux/-/merge_requests/306
There's already a bug for Ubuntu: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1905975

mauriciovasquezbernal commented on src/core/load-fragment.c:
```diff
@@ -3509,6 +3509,73 @@ int config_parse_restrict_namespaces(
+                void *userdata) {
+        _cleanup_strv_free_ char **n = NULL;
+        size_t nlen = 0, nbufsize = 0;
+        char*** restrict_fs = data;
```
```suggestion
        char ***restrict_fs = data;
```

?

mauriciovasquezbernal commented on man/systemd.exec.xml:
```diff
@@ -1699,6 +1699,25 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+        <listitem><para>Restricts the set of filesystems processes of this unit can open files on. Takes a space-separated
+        list of filesystem names. Any filesystem listed is made accessible to the unit's processes, access to filesystem
+        types not listed is prohibited (allow-listing). If the empty string is assigned, access to filesystems is not
+        restricted. This option may appear more than once, in which case the namespace
+        types are merged by <constant>OR</constant>.</para>
```
Namespace types?

mauriciovasquezbernal commented on src/core/bpf/restrict_fs/restrict-fs.c:
```diff
@@ -0,0 +1,59 @@
+int BPF_PROG(restrict_filesystems, struct file *file, int ret)
+{
+        unsigned long magic_number;
+        unsigned long tmpfs_magic;
```
Unused variable, right?

iaguis commented:
I've addressed (I think) all review comments. Some high level changes:

* `RestrictFileSystems=` accepts either an allow list or a deny list similar to `SystemCallFilter`.
* I use gperf to generate the filesystem database.
* I've added some filesystem groups as suggested and used those in existing functions.
    * I've added a `systemd-analyze filesystems` command to discover those groups.
* I don't generate the filesystem list automatically by parsing `magic.h` because filesystem names are kinda arbitrary, but I've added a script that checks all filesystems defined in `magic.h` are present in the database.

Also, I was wondering if we should split the @wat-ze-hex's commits to a separate PR now that there are 3 PRs using it (#17655, #18385, and this one). It might make things easier to merge.

wat-ze-hex commented:
@iaguis 

I addressed comments [[0]](https://github.com/systemd/systemd/pull/17655#issuecomment-771736857) [[1]](https://github.com/systemd/systemd/pull/17655#issuecomment-772013678) of @anakryiko and @mauriciovasquezbernal and switched to BPF skeleton instead of hexdump generation, wdyt? Attention worthy commits from the most recent push are https://github.com/systemd/systemd/pull/17655/commits/bc724de2eb37f0908ef201b5ca88c88dc57f1da4, https://github.com/systemd/systemd/pull/17655/commits/c1890737f1e8f9fa077431652053e554e657fefc.

Also BPF skeleton serves pretty much all needs addressed previously by custom `bpf_object` helpers [at least for #17655 feature] so I removed those sources altogether from my PR.

I like the idea about separating framework part into its own PR. Please give me couple of days to study your PR and #18385 and wrap my mind around BPF helpers we might need for all three.

mauriciovasquezbernal commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,375 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to find LSM BPF program '%s': %m", LSM_PROG_SECTION_NAME);
+
+        link = bpf_program__attach_lsm(prog);
```
This will attach the program to the LSM hook in all the cases even if the `RestrictFileSystems` feature is not used at all. Is that possible to implement a conditional logic that only (loads? and) attaches the program when that property is needed?

wat-ze-hex replied:
`bpf_link` is leaked here: memory allocated in https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L9893 is not freed.
Also link fd is not closed and link stays attached after `lsm_bpf_setup` exits.
`bpf_link__destroy` should be used to close fd and free memory, e.g. in cleanup function https://github.com/systemd/systemd/pull/17655/commits/8326f7cc9640c5dd4a61335291fa9b292a893ae6

Also for bpf_link to stay attached, it must be stored. `Manager` seems to be a proper place for it.

mauriciovasquezbernal commented on README:
```diff
@@ -106,6 +111,13 @@ REQUIREMENTS:
+          CONFIG_BPF_SYSCALL
+          CONFIG_BPF_LSM
+          CONFIG_DEBUG_INFO_BTF
+          CONFIG_LSM="...,bpf"
```
This can be overwritten by the `lsm=` kernel boot parameter. Maybe the following makes sense? 

```suggestion
          CONFIG_LSM="...,bpf" or boot the kernel with lsm="...,bpf". 
```

wat-ze-hex commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,375 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed create systemd bpffs directory: %m");
+
+        r = bpf_obj_pin(outer_map_fd, LSM_MAP_BPFFS_PATH);
```
Pinning step can be dropped altogether if `outer_map_fd` is stored. Alternatively bpf skeleton generated object provide `bpf_map` pointer. Similar to `bpf_link`, `Manager` looks like a proper place to store either object or fd.
IMO using bpffs questionable in this context, since it's a security feature and with bpffs too much container internals are shared with host. Also bpffs may not be mounted.

iaguis replied:
Make sense, I'll give it a go and change thus to use the BPF skeleton approach.

wat-ze-hex reviewed (commented):
@iaguis 
my 2c, more comments will follow

bluca commented:
> > > Note the BPF LSM hook is very new and is not enabled in current distro kernels AFAIK.
> > 
> > 
> > Hmm, pity. Is this tracked anywhere? Did anyone ever file a bug against let's say Fedora asking for this to be added yet?
> 
> I'll start with Debian: https://salsa.debian.org/kernel-team/linux/-/merge_requests/306
> There's already a bug for Ubuntu: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1905975

FYI, the BPF LSM will be available (but disabled by default, kernel command line / sysfs to turn it on at rutnime) in Debian 11: https://salsa.debian.org/kernel-team/linux/-/merge_requests/308/diffs

iaguis commented:
BPF LSM has been added and enabled by default in Arch it seems: https://github.com/archlinux/svntogit-packages/commit/69cb8c2d2884181e799e67b09d67fcf7944d8408

poettering commented:
Any chance you can rebase now that #17655 has been merged? Should be an easy merge then

poettering commented:
@iaguis ping?

iaguis commented:
Sorry for dropping this. I plan on reworking this PR next week.

bluca commented on src/basic/filesystems.h:
```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "stat-util.h"
```
nit: sort alphabetically

bluca commented on src/basic/filesystems.h:
```diff
@@ -0,0 +1,34 @@
+        FILESYSTEM_SET_NETWORK,
+        FILESYSTEM_SET_TEMPORARY,
+        FILESYSTEM_SET_KNOWN,
+        _FILESYSTEM_SET_MAX
```
`_FILESYSTEM_SET_INVALID = -EINVAL,`

bluca commented on src/basic/filesystems-gperf.gperf:
```diff
@@ -0,0 +1,111 @@
+
+#include "filesystems.h"
+#include "missing_magic.h"
+#include "missing_fs.h"
```
sort alphabetically

bluca commented on src/basic/filesystems.h:
```diff
@@ -0,0 +1,34 @@
+#include "nulstr-util.h"
+#include "string-util.h"
+
+#define FILESYSTEM_MAGIC_MAX 10
```
isn't the enum enough?

iaguis replied:
This is about the number of magic numbers per filesystem type in the `filesystems-gperf.gperf` table. For example for the "minix" name there are 5 magic numbers associated with it.

bluca commented on src/core/meson.build:
```diff
@@ -132,6 +132,8 @@ libcore_sources = '''
 subdir('bpf/socket_bind')
 if conf.get('BPF_FRAMEWORK') == 1
         libcore_sources += [socket_bind_skel_h]
+        subdir('bpf/restrict_fs')
```
this commit "bpf: add meson build rules for the restrict_fs program" can be squashed in the previous one "bpf: add restrict_fs BPF program", it doesn't make much sense standalone?

bluca commented on src/basic/cgroup-util.c:
```diff
@@ -1318,6 +1318,28 @@ int cg_pid_get_machine_name(pid_t pid, char **machine) {
+        if (name_to_handle_at(AT_FDCWD, path, h, &mnt_id, 0) < 0)
+                return -errno;
+
+        if (ret)
```
There's both `assert(ret)` and `if (ret)` - you need only one. If it makes sense to call this without getting the ID back, remove the `assert`, otherwise the `if`.

bluca commented on src/basic/meson.build:
```diff
@@ -366,6 +366,34 @@ run_target(
+filesystem_includes = ['linux/magic.h',
+                       'linux/gfs2_ondisk.h']
+
+check_filesystems = find_program('check-filesystems.sh')
```
these scripts and files were added by a previous commit - the meson integration should be in the same commit

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
```
please sort the includes - system headers in the first block, internal headers in the second block. Both blocks sorted alphabetically.

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+                return log_error_errno(errno, "Failed to open BPF object: %m");
+
+        /* TODO Maybe choose a number based on runtime information? */
+        r = sym_bpf_map__resize(obj->maps.cgroup_hash, 2048);
```
make the magic number a \#define at the top please

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+                return log_error_errno(r, "Failed to load BPF object");
+
+        /* Close dummy map */
+        inner_map_fd = safe_close(inner_map_fd);
```
this happens automatically, no?

iaguis replied:
Right

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+        static int cached_use = -1;
+        int r;
+
+        if (cached_use < 0) {
```
nit: indentation in the whole function can be reduced by switching this around, ie:

```
if (cached_use >= 0)
  return cached_use;
```

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+
+        if (r == 0) {
+                log_info_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                         "Not running with unified cgroup hierarchy, LSM BPF is not supported");
```
nit: indentation when breaking function calls into multiple lines, has the parameters aligned (ie: move this further right)

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+        if (r < 0)
+                return log_info_errno(r, "Failed to load BPF object: %m");
+
+        return bpf_can_link_lsm_program(obj->progs.restrict_filesystems);
```
is it intentional that these 2 calls above don't update the cache flag?

iaguis replied:
Oops, I missed those. Thanks!

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+        struct restrict_fs_bpf *obj = NULL;
+        _cleanup_(bpf_link_freep) struct bpf_link *link = NULL;
+        int r;
+
```
`assert(m)`

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+
+                log_unit_debug(u, "Restricting filesystem access to '%s'", fs);
+
+                for (i=0; i<FILESYSTEM_MAGIC_MAX; i++) {
```
move the declaration of `i` here - also whitespace between symbols

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,335 @@
+
+int bpf_map_restrict_fs_fd(Unit *unit) {
+        assert(unit);
+
```
`assert(unit->manager)`

bluca commented on src/core/bpf-lsm.c:
```diff
@@ -333,3 +333,71 @@ void lsm_bpf_destroy(struct restrict_fs_bpf *prog) {
+                                default:
+                                        return r;
+                                }
+                } else {
```
nit: no need for brackets for the `else` case

bluca commented on .mkosi/mkosi.debian:
```diff
@@ -46,6 +49,8 @@ BuildPackages=
         libxkbcommon-dev
         libzstd-dev
+        llvm
+        m4
```
why m4?

bluca commented on src/basic/check-filesystems.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -eu
```
`# SPDX-License-Identifier: LGPL-2.1-or-later`
`set -o pipefail`

bluca commented on src/basic/filesystems-gperf.gperf:
```diff
@@ -0,0 +1,111 @@
+%{
```
`{# SPDX-License-Identifier: LGPL-2.1-or-later #}`

bluca commented on test/fuzz/fuzz-unit-file/directives.mount:
```diff
@@ -143,6 +143,7 @@ ReadWritePaths=
 RemoveIPC=
 RestartKillSignal=
 RestrictAddressFamilies=
+RestrictFileSystems=
```
merge this commit that adds to test/fuzz into the commit that adds the directive - otherwise those tests that check for this will fail until this commit

bluca commented on src/systemctl/systemctl-show.c:
```diff
@@ -1047,7 +1047,7 @@ static int print_property(const char *name, const char *expected_value, sd_bus_m
                         return 1;
 
-                } else if (STR_IN_SET(name, "SystemCallFilter", "SystemCallLog", "RestrictAddressFamilies")) {
+                } else if (STR_IN_SET(name, "SystemCallFilter", "SystemCallLog", "RestrictAddressFamilies", "RestrictFileSystems")) {
```
this commit "core: show RestrictFileSystems= property" should be folded in the commit that adds the directive

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+        _cleanup_set_free_ Set *filesystems = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+
```
`assert(ret)`

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+
+static void filesystem_set_remove(Set *s, const FilesystemSet *set) {
+        const char *filesystem;
+
```
```
if (!set)
 return
```

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+
+static void dump_filesystem(const FilesystemSet *set) {
+        const char *filesystem;
+
```
assert or return on null (set)

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+
+                k = load_available_kernel_filesystems(&kernel);
+
+                for (i = 0; i < _FILESYSTEM_SET_MAX; i++) {
```
declare `i` in the for

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
 }
 #endif
 
+#if HAVE_LIBBPF
```
there's no usage of libbpf here? So you can reduce the scope of the `ifdef` a lot, maybe just an early return in `dump_filesystems`

bluca commented on src/analyze/analyze.c:
```diff
@@ -1802,6 +1804,169 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+
+        /* Let's read the available filesystems */
+
+        f = fopen("/proc/filesystems", "re");
```
you should be able to use `read_virtual_file` instead of re-implementing it

iaguis replied:
`read_virtual_file` doesn't read line by line, which is what I need here. Do you think it's worth it using it and reading the resulting string line by line to get the filesystems?

bluca replied:
I'd say so, there are annoyances when reading kernel API filesystems that we want to generally abstract away in one place

bluca reviewed (changes requested):
Looks great! A few nits here and there, but nothing major

iaguis commented:
Thanks a lot for the review! I think I've addressed all comments.

Also, I notice some unit tests failed because they were calling `manager_new()` but not `lsm_bpf_setup()` because it's called in the `main()` function. I think it makes sense `lsm_bpf_setup()` goes in `manager_new()` so I moved it over and this fixes the tests.

PTAL.

bluca commented on src/core/manager.c:
```diff
@@ -905,6 +905,14 @@ int manager_new(UnitFileScope scope, ManagerTestRunFlags test_run_flags, Manager
         }
 
+#if HAVE_LIBBPF
+        if (MANAGER_IS_SYSTEM(m) && lsm_bpf_supported()) {
```
this block should probably be in the next if? We probably don't want to run this in systemd-analyze

iaguis replied:
I'm not very familiar with this code. Do you mean above in the `else` block? Calls in `src/analyze` set the `MANAGER_TEST_RUN_MINIMAL` so I think that makes sense but maybe I'm missing something.

bluca replied:
yeah they both check `test_run_flags`, the above block also looks fine to place it

bluca reviewed (approved):
LGTM - others had comments too, so leaving open for a few days to give them opportunity to check again

iaguis commented:
Friendly ping :hugs:

poettering commented:
(sorry, been on vacation, still catching up with reviews)

needs rebase

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -345,3 +345,70 @@ void lsm_bpf_destroy(struct restrict_fs_bpf *prog) {
+                 * we want to allow it, then remove it from the list. */
+                if (!(flags & FILESYSTEM_PARSE_INVERT) == !!(flags & FILESYSTEM_PARSE_ALLOW_LIST)) {
+                        SET_FOREACH(fs, filesystems) {
+                                if (streq(fs, name)) {
```
```c
if (set_contains(filesystems, name))
        return 0;
```

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -345,3 +345,70 @@ void lsm_bpf_destroy(struct restrict_fs_bpf *prog) {
+                } else
+                        SET_FOREACH(fs, filesystems) {
+                                if (streq(fs, name)) {
+                                        free(set_remove(filesystems, fs));
```
hmm, you can replace this loop with `free(set_remove(filesystems, name))`? what am i missing?

poettering commented on src/core/load-fragment.c:
```diff
@@ -3539,6 +3540,80 @@ int config_parse_restrict_namespaces(
+        }
+
+        if (!c->restrict_filesystems) {
+                c->restrict_filesystems = set_new(NULL);
```
so you are usual "trivial_hash_ops" here, i.e. the set object is keyed by hashed pointer values, not by the strings they point to. This is weird, what's the point? You want to pass `string_hash_ops` or so here.

But actually, don't allocate this here at all. The nice thing about set_put_strdup() is that it allocates the set if not allocated yet, i.e. lazily, and implicitly uses the string hash ops for that with destructors. so you can shorten this quite a bit

poettering commented on src/core/dbus-execute.c:
```diff
@@ -689,6 +690,53 @@ static int property_get_input_data(
+                r = strv_extend(&l, f);
+                if (r < 0)
+                        return -ENOMEM;
+        }
```
set_get_strv()

poettering commented on src/core/dbus-execute.c:
```diff
@@ -1882,6 +1931,71 @@ int bus_exec_context_set_transient_property(
+                        if (!c->restrict_filesystems) {
+                                c->restrict_filesystems = set_new(NULL);
+                                if (!c->restrict_filesystems)
+                                        return log_oom();
```
as above, drop this, you can allocate the set on-the-fly via set_put_strdup()

poettering commented on src/analyze/analyze.c:
```diff
@@ -1804,6 +1809,172 @@ static int dump_syscall_filters(int argc, char *argv[], void *userdata) {
+
+                k = load_available_kernel_filesystems(&kernel);
+
+                for (int i = 0; i < _FILESYSTEM_SET_MAX; i++) {
```
instead of "int" use the enum type you defined

anakryiko commented on src/core/bpf/restrict_fs/restrict-fs.bpf.c:
```diff
@@ -0,0 +1,72 @@
+        return 0;
+}
+
+char _license[] SEC("license") = "GPL";
```
this can be const, I think, but it shouldn't be marked static, otherwise compiler will just remove it. It needs to be preserved for libbpf and kernel. Maybe static + `__attribute__((unused))` would work, but no one is doing it that way, as far as I know. Libbpf just expects that there will be ELF section "license" with contents being the license for all BPF programs in this file.

iaguis replied:
How about leaving it `const char[]`?

poettering replied:
so far we had the rule that symbols that have no prototype in a header file should be static. i.e. the "static" is not about whether the stuff should appear in the final image or not, but simply whether it shall be referenced by its symbol name. But here the symbol name doesn't really matter, you could name the symbol "quux" if you'd like with no ill effects. What matters is that the stuff us marked with `__attribute__((__used__))` and has the section attribute too. Turns out he SEC() macro already adds both.

We use the section export stuff at various othe rplaces, for example in the sd-boot "SBAT" section. We always specify "static" there, hence it should be safe to do so here too.

anakryiko commented on src/core/bpf/restrict_fs/restrict-fs.bpf.c:
```diff
@@ -0,0 +1,72 @@
+
+SEC("lsm/file_open")
+int BPF_PROG(restrict_filesystems, struct file *file, int ret)
+{
```
I might be misremembering (sorry, too lazy to find it in previous PRs), but I think we discussed coding style of .bpf.c files and decided to keep them in the kernel coding style w/ mostly C89 standard, consistent with all other open-source cases (like libbpf-tools, kernel tests themselves, a bunch of other BPF-based projects, etc). It's nice to stay consistent for BPF code and make it easier for contributors to read and write.

iaguis replied:
Found the reference: https://github.com/systemd/systemd/pull/17655#discussion_r612696922. I don't have a strong opinion on this.

iaguis commented on src/basic/missing_fs.h:
```diff
@@ -65,3 +65,43 @@
+/* Not exposed yet. Defined at fs/exfat/exfat_fs.h */
+#ifndef EXFAT_SUPER_MAGIC
+#define EXFAT_SUPER_MAGIC 0x2011BAB0UL
+#endif
```
I followed the convention of defining non-exposed FS in missing_fs.h and exposed but maybe not available in missing_magic.h. Only CIFS_MAGIC_NUMBER was present in missing_fs.h before my changes so it's not a strong convention and I see there's one non-exposed FS in missing_magic.h.

I agree it makes more sense to have all magic numbers in missing_magic.h, I'll move them over.

iaguis commented on src/basic/filesystems.c:
```diff
@@ -0,0 +1,117 @@
+                .name = "@basic-api",
+                .help = "Basic filesystem API",
+                .value =
+                "autofs\0"
```
I think I'll just remove it then.

iaguis commented on src/basic/filesystems.c:
```diff
@@ -0,0 +1,117 @@
+                "mqueue\0"
+                "proc\0"
+                "pstore\0"
+                "ramfs\0"
```
OK.

I copied this list from https://github.com/systemd/systemd/blob/v249/src/basic/mountpoint-util.c#L374-L396, should we remove it from there too? From what I can see this is used to check if a fs type is blockdev backed so maybe it's fine to keep it there.

I'll remove it from this list.

poettering replied:
hmm, it might indeed be misplaced there too. or maybe that function should be renamed. something indeed is off

iaguis commented on src/basic/filesystems.c:
```diff
@@ -0,0 +1,117 @@
+                "sysfs\0"
+                "tmpfs\0"
+                "tracefs\0"
+        },
```
how about @extended-api for this second group?

poettering replied:
yeah, or "auxiliary-api" or so ?

iaguis commented on src/basic/filesystems.c:
```diff
@@ -0,0 +1,117 @@
+                .help = "Common block device filesystems",
+                .value =
+                "btrfs\0"
+                "ext3\0"
```
Makes sense. I can't think of any other block fs for that group, any ideas?

poettering replied:
maybe ext3, too? older minix?

iaguis commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,347 @@
+                if (r < 0)
+                        return 0;
+
+                if (strstr(word, "bpf")) {
```
oops

poettering commented on src/core/execute.c:
```diff
@@ -4659,6 +4703,15 @@ static int exec_child(
 #endif
+
+#if HAVE_LIBBPF
+                r = apply_restrict_filesystems(unit, context);
```
hmm, the other bpf hooks are applied via the cgroups logic, i.e. see how CGROUP_CONTROLLER_BPF_SOCKET_BIND works. I think that's a better approach, since we want to install the bpf program when realizing the cgroup and not when forking off a process, since we might fork off many processes, and we shouldn't update the cgroup each time.

iaguis replied:
The BPF program is installed only once when creating the system manager. It's true that with current code we're updating the global BPF map with every process we fork off so that's not ideal.

If we move the code to the cgroups logic we'd need to allow some filesystems so things are correctly set up like tmpfs for STDIN or cgroup to attach the cgroup. And if we allow them temporarily and removed them later we're doing a map operation anyway so we might as well not restrict until the last moment.

poettering replied:
oh, of course, that makes sense. (might be cool to add a comment to the definition of CGroupController enum that quickly mentions this before the next person looking at this code gets confused)

iaguis commented:
I think I've addressed everything. PTAL.

bluca commented on man/systemd.exec.xml:
```diff
@@ -3834,6 +3834,11 @@ StandardInputData=SWNrIHNpdHplIGRhIHVuJyBlc3NlIEtsb3BzLAp1ZmYgZWVtYWwga2xvcHAncy
+          <row>
+            <entry>245</entry>
+            <entry><constant>EXIT_BPF</constant></entry>
+            <entry>Failed to set apply BPF restrictions. See <varname>RestrictFileSystems=</varname> above.</entry>
```
"Failed to apply BPF..."

iaguis commented on src/core/execute.c:
```diff
@@ -4943,6 +4996,8 @@ void exec_context_done(ExecContext *c) {
         c->apparmor_profile = mfree(c->apparmor_profile);
         c->smack_process_label = mfree(c->smack_process_label);
 
+        c->restrict_filesystems = set_free(c->restrict_filesystems);
```
After removing the `set_new()` code as suggested by https://github.com/systemd/systemd/pull/18145#discussion_r690350651 and [changing bpf_lsm_parse_filesystems to initialize the set in there](https://github.com/systemd/systemd/pull/18145/files#diff-d37918508c319e9033e0f8281e72a1bd7f415d55377b42b3d3c7412c1a7b00c2R23) I'm seeing crashes here:

```gdb
Core was generated by `/sbin/init'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007ff52744101b in kill () from /usr/lib/libc.so.6
[Current thread is 1 (LWP 779)]
(gdb) bt
#0  0x00007ff52744101b in kill () from /usr/lib/libc.so.6
#1  0x00005563d444f854 in crash (sig=6) at ../src/core/main.c:262
#2  <signal handler called>
#3  0x00007ff527440d22 in raise () from /usr/lib/libc.so.6
#4  0x00007ff52742a862 in abort () from /usr/lib/libc.so.6
#5  0x00007ff527482d28 in __libc_message () from /usr/lib/libc.so.6
#6  0x00007ff52748a92a in malloc_printerr () from /usr/lib/libc.so.6
#7  0x00007ff52748b80c in unlink_chunk.constprop () from /usr/lib/libc.so.6
#8  0x00007ff52748c07b in _int_free () from /usr/lib/libc.so.6
#9  0x00007ff52748f9e8 in free () from /usr/lib/libc.so.6
#10 0x00007ff5278b7420 in _hashmap_clear (h=0x5563d5bd5f78, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:924
#11 0x00007ff5278b730d in _hashmap_free (h=0x5563d5bd5f78, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:888
#12 0x00005563d453b5cb in hashmap_free (h=0x5563d5bd5f78) at ../src/basic/hashmap.h:102
#13 0x00005563d453f006 in unit_clear_dependencies (u=0x5563d5c39c70) at ../src/core/unit.c:558
#14 0x00005563d453f7c0 in unit_free (u=0x5563d5c39c70) at ../src/core/unit.c:698
#15 0x00005563d44d2a78 in manager_clear_jobs_and_units (m=0x5563d5b04c30) at ../src/core/manager.c:1437
#16 0x00005563d44da840 in manager_reload (m=0x5563d5b04c30) at ../src/core/manager.c:3324
#17 0x00005563d4458ae8 in invoke_main_loop (m=0x5563d5b04c30, saved_rlimit_nofile=0x7fff74b947b0, saved_rlimit_memlock=0x7fff74b947c0, ret_reexecute=0x7fff74b946db,
    ret_retval=0x7fff74b946e0, ret_shutdown_verb=0x7fff74b94730, ret_fds=0x7fff74b94748, ret_switch_root_dir=0x7fff74b94720, ret_switch_root_init=0x7fff74b94728,
    ret_error_message=0x7fff74b94738) at ../src/core/main.c:1966
#18 0x00005563d445c5c8 in main (argc=1, argv=0x7fff74b949c8) at ../src/core/main.c:2942
```

I'm probably missing something obvious but I can't figure out why this happens.

iaguis replied:
Sorry, pasted the wrong backtrace, this is the right one:

```
#0  0x00007fcb3c0a501b in kill () from /usr/lib/libc.so.6
#1  0x000055a50e8e9854 in crash (sig=6) at ../src/core/main.c:262
#2  <signal handler called>
#3  0x00007fcb3c0a4d22 in raise () from /usr/lib/libc.so.6
#4  0x00007fcb3c08e862 in abort () from /usr/lib/libc.so.6
#5  0x00007fcb3c0e6d28 in __libc_message () from /usr/lib/libc.so.6
#6  0x00007fcb3c0ee92a in malloc_printerr () from /usr/lib/libc.so.6
#7  0x00007fcb3c0f0408 in _int_free () from /usr/lib/libc.so.6
#8  0x00007fcb3c0f39e8 in free () from /usr/lib/libc.so.6
#9  0x00007fcb3c51965c in __unique_prefix_static_free_wrapper5 (a=0x55a50f18fb40) at ../src/basic/hash-funcs.c:13
#10 0x00007fcb3c51b3dd in _hashmap_clear (h=0x55a50f1f68a0, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:916
#11 0x00007fcb3c51b30d in _hashmap_free (h=0x55a50f1f68a0, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:888
#12 0x000055a50e918cb6 in set_free (s=0x55a50f1f68a0) at ../src/basic/set.h:20
#13 0x000055a50e9335ce in exec_context_done (c=0x55a50f192818) at ../src/core/execute.c:4999
#14 0x000055a50e9d94d3 in unit_done (u=0x55a50f1922e0) at ../src/core/unit.c:643
#15 0x000055a50e9d9595 in unit_free (u=0x55a50f1922e0) at ../src/core/unit.c:664
#16 0x000055a50e96ac40 in manager_dispatch_cleanup_queue (m=0x55a50f178c30) at ../src/core/manager.c:1170
#17 0x000055a50e9732f1 in manager_loop (m=0x55a50f178c30) at ../src/core/manager.c:3016
#18 0x000055a50e8f28d8 in invoke_main_loop (m=0x55a50f178c30, saved_rlimit_nofile=0x7ffcf08c9590, saved_rlimit_memlock=0x7ffcf08c95a0, ret_reexecute=0x7ffcf08c94bb,
    ret_retval=0x7ffcf08c94c0, ret_shutdown_verb=0x7ffcf08c9510, ret_fds=0x7ffcf08c9528, ret_switch_root_dir=0x7ffcf08c9500, ret_switch_root_init=0x7ffcf08c9508,
    ret_error_message=0x7ffcf08c9518) at ../src/core/main.c:1933
#19 0x000055a50e8f65c8 in main (argc=1, argv=0x7ffcf08c97a8) at ../src/core/main.c:2942
```

iaguis replied:
The cause was https://github.com/systemd/systemd/pull/18145#discussion_r692757285

iaguis commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,404 @@
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup ID: %m");
```
I'm seeing a bunch of messages with this error:

```
[vagrant@archlinux ~]$ sudo journalctl -b -3 | grep "Failed to get cgroup ID"
Aug 19 14:54:56 archlinux systemd[1]: systemd-homed-activate.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: systemd-coredump.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: git.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: vagrant.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: run-user-1000.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: dbus.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-networkd-wait-online.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-update-done.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: ldconfig.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-hwdb-update.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-journal-catalog-update.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-tmpfiles-setup.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-networkd.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-sysctl.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-tmpfiles-setup-dev.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-sysusers.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-remount-fs.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-reboot.service: Failed to get cgroup ID: No such file or directory
```

It seems name_to_handle_at() fails with ENOENT so the cgroup path might already be gone. I wonder if we should make this fail silently in this case.

iaguis replied:
Maybe a better way is to save the `cgroup_id` in `struct Unit` next to the `cgroup_path` and then this can be used by the `bpf_restrict_filesystems` and `cleanup_lsm_bpf()`. Wdyt?

iaguis replied:
Added a couple of commits implementing this. I can squash the bpf-lsm commit if this approach is ok.

mauriciovasquezbernal commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,404 @@
+                /* If we previously wanted to forbid access to a filesystem and now
+                 * we want to allow it, then remove it from the list. */
+                if (!(flags & FILESYSTEM_PARSE_INVERT) == !!(flags & FILESYSTEM_PARSE_ALLOW_LIST)) {
+                        if (set_contains(*filesystems, name)) /* Already present, ignoring */
+                                return 0;
```
nit: `set_put_strdup` ends calling `_set_put_strdup_full` that [performs](https://github.com/systemd/systemd/blob/7b58fcdd705aad449c60ae7d23a6f1d5a79bd737/src/basic/hashmap.c#L1855) this same check. So I think it can be avoided.

krnowak commented on src/core/dbus-execute.c:
```diff
@@ -689,6 +690,47 @@ static int property_get_input_data(
+                sd_bus_error *error) {
+
+        ExecContext *c = userdata;
+        _cleanup_strv_free_ char **l = NULL;
```
This should be `_cleanup_free` otherwise you'll free the set keys, those are not copied when calling `set_get_strv`.

iaguis replied:
Ah, that makes sense. Great catch! Thank you!

poettering commented on src/basic/mountpoint-util.c:
```diff
@@ -351,48 +353,30 @@ bool fstype_is_network(const char *fstype) {
+        if (nulstr_contains(filesystem_sets[FILESYSTEM_SET_BASIC_API].value, fstype) ||
+            nulstr_contains(filesystem_sets[FILESYSTEM_SET_AUXILIARY_API].value, fstype) ||
+            nulstr_contains(filesystem_sets[FILESYSTEM_SET_PRIVILEGED_API].value, fstype) ||
+            nulstr_contains(filesystem_sets[FILESYSTEM_SET_TEMPORARY].value, fstype))
```
could also be:

```c
FilesystemSet *fs;

FOREACH_POINTER(fs, 
        filesystem_sets + FILESYSTEM_SET_BASIC_API, 
        filesystem_sets + FILESYSTEM_SET_AUXILIARY_API, 
        filesystem_sets + FILESYSTEM_SET_PRIVILEGED_API, 
        filesystem_sets + FILESYSTEM_SET_TEMPORARY_API)
    if (nulstr_contains(fs->value, fstype))
            return true;
return false;
```
but i guess this doesn't matter

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1367,6 +1367,32 @@ int cg_pid_get_machine_name(pid_t pid, char **machine) {
+
+        /* This is cgroupfs so we know the size of the handle, thus no need to loop around like
+         * name_to_handle_at_loop() does in mountpoint-util.c */
+        if (name_to_handle_at(AT_FDCWD, path, h, &mnt_id, 0) < 0)
```
why not pass `&buf.f_handle` as third arg here? requires no ugly casts then.

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1367,6 +1367,32 @@ int cg_pid_get_machine_name(pid_t pid, char **machine) {
+        if (name_to_handle_at(AT_FDCWD, path, h, &mnt_id, 0) < 0)
+                return -errno;
+
+        *ret = *(uint64_t *) h->f_handle;
```
and here: 
```c
*ret = *(uint64_t*) buf.f_handle.f_handle;
```

And then drop `h`?

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1367,6 +1367,32 @@ int cg_pid_get_machine_name(pid_t pid, char **machine) {
+        assert(path);
+        assert(ret);
+
+        struct {
```
union, not struct

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,347 @@
+                if (r == -ENOMEM)
+                        return log_oom();
+                if (r < 0)
+                        return 0;
```
probably should debug log here that there was a parse error we ignore, i.e.

```c
if (r < 0) {
        log_debug_errno("Failed to parse /sys/kernel/security/lsm, ignoring: %m");
        return 0;
}
```

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,347 @@
+        r = read_one_line_file("/sys/kernel/security/lsm", &lsm_list);
+        if (r < 0) {
+               if (errno != ENOENT)
+                       log_debug_errno(r, "Failed to read /sys/kernel/security/lsm: %m");
```
maybe suffix the log message with `, ignoring: %m` 

we usually do this for log messages we consider something to ignore

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,347 @@
+        r = prepare_restrict_fs_bpf(&obj);
+        if (r < 0) {
+                return supported = 0;
+        }
```
doesn't matter, but no {} around single-line if blocks according to our coding style

poettering commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,347 @@
+}
+
+int bpf_map_restrict_fs_fd(Unit *unit) {
+        return -1;
```
maybe -ENOMEDIUM as above?

or -EBADF?

poettering commented on src/core/cgroup.h:
```diff
@@ -3,6 +3,9 @@
 #include <stdbool.h>
 
+#if HAVE_LIBBPF
+#include "bpf-lsm.h"
```
hmm, this header can be imported always, no? i.e. it's written to also be clean to import if HAVE_LIBBPF is 0, no?

I'd generally try to minimize ifdeffery in consuming files, and place it in implementing files instead. Thus, i'd just include bpf-lsm.h here unconditionally without ifdeffery

poettering commented on src/core/execute.c:
```diff
@@ -41,6 +41,9 @@
 #include "barrier.h"
+#if HAVE_LIBBPF
+#include "bpf-lsm.h"
+#endif
```
as above

poettering commented on src/core/bpf-lsm.h:
```diff
@@ -0,0 +1,16 @@
+int lsm_bpf_supported(void);
+int lsm_bpf_setup(Manager *m);
+int bpf_restrict_filesystems(Unit *u, const Set *filesystems, bool allow_list);
+int cleanup_lsm_bpf(const Unit *u);
```
hmm, most other symbols here use `lsm_bpf` as prefix, not suffix. do so here too? (and in fact for al symbols?

the ones that take a Unit* as first arg might even use `lsm_bpf_unit_` as prefix?

iaguis replied:
Makes sense, I'll make them all consistent.

poettering commented on src/core/dbus-execute.c:
```diff
@@ -1882,6 +1925,67 @@ int bus_exec_context_set_transient_property(
         if (streq(name, "RestrictNamespaces"))
                 return bus_set_transient_namespace_flag(u, name, &c->restrict_namespaces, message, flags, error);
 
+#if HAVE_LIBBPF
```
this should not be ifdeffed.

So far we accept all props in all cases, we just don't appy them if we lack the support for them. Hence here too: please parse this stuff, and make it available via the bus props. so that any version of systemd compiled from the same sources accepts and advertises the same set of props with the right data, even if it doesn't actually apply the settings.

poettering commented on src/core/cgroup.c:
```diff
@@ -2074,6 +2076,16 @@ static int unit_update_cgroup(
                 return log_unit_error_errno(u, r, "Failed to create cgroup %s: %m", empty_to_root(u->cgroup_path));
         created = r;
 
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_full_path);
```
this is a cgroup2 concept.  maybe conditionalize the cg_get_path already under `cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER)`

poettering commented on src/core/unit.h:
```diff
@@ -293,6 +293,7 @@ typedef struct Unit {
 
         /* Counterparts in the cgroup filesystem */
         char *cgroup_path;
+        uint64_t cgroup_id;
```
hmm, is the cgroup id 0 an invalid cgroup id? If not we should probably initialize this to UINT64_MAX or so, as a marker for "doesn't apply/don't know" instead of zero.

iaguis replied:
From a cursory look at the kernel, [cgroup IDs](https://github.com/torvalds/linux/blob/v5.14/include/linux/cgroup.h#L310-L313) are [inodes](https://github.com/torvalds/linux/blob/v5.14/include/linux/kernfs.h#L150-L154) which AFAIK can't be 0. I also found some [other reference](https://github.com/torvalds/linux/blob/v5.14/tools/testing/selftests/bpf/cgroup_helpers.c#L242-L243) saying that 0 is an invalid cgroup id, but it's just a comment.

Given this I think 0 is fine but I could change it to UINT64_MAX to be safe. Wdyt?

bluca replied:
we usually use _MAX for "invalid"

poettering reviewed (changes requested):
Looks excellent. Just some super minor stuff. 

Excellent work! Thanks so much!

And very sorry my reviews took so long.

iaguis commented:
Thanks a lot for the review!

bluca commented on src/basic/generate-filesystem-list.py:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python3
+import sys
```
SPDX line is missing

bluca commented on src/core/bpf/restrict_fs/restrict-fs.bpf.c:
```diff
@@ -0,0 +1,72 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
Why is this GPL-2.0-only? We cannot accept sources that are not compatible with LGPL-2.1-or-later, as it means the license of the entire project is affected

bluca replied:
is this source copied/derived from an existing GPL-2.0-only project?

iaguis replied:
It's not copied/derived.

Originally this was set as GPL-2.0 only because the kernel enforces the "license" section in the ELF binary to be "GPL". However I see in https://github.com/systemd/systemd/blob/v249/src/core/bpf/socket_bind/socket-bind.bpf.c#L1-L7 there's a note explaining that LGPL-2.1-or-later downgrades to GPL so it's fine to add it as LGPL-2.1-or-later in the SPDX header.

IANAL but I guess I can add a similar comment here.

bluca replied:
Yes, the note is correct, please do the same thing here

poettering commented on src/core/dbus-execute.c:
```diff
@@ -681,6 +682,47 @@ static int property_get_input_data(
+        return sd_bus_message_close_container(reply);
+}
+
+
```
spurious double empty line

poettering commented on src/core/dbus-execute.c:
```diff
@@ -1875,6 +1918,65 @@ int bus_exec_context_set_transient_property(
+
+                        if (!c->restrict_filesystems) {
+                                c->restrict_filesystems_allow_list = allow_list;
+                        }
```
doesn't matter, but as per coding style, no {} around singl-line if blocks

poettering reviewed (commented):
two trivialities

poettering commented:
excellent work. found to trivial issues, but they don#t really matter. setting green label.  (i.e would be great if you'd fix those too, but we'd merge even without)

iaguis commented:
Fixed those two nits :)

poettering commented:
was about to merge, but needs a rebase i fear.

poettering commented:
lgtm, thanks!
